### PR TITLE
Flow-based infrastructure reconciliation without Terraformer

### LINKED ISSUE #457: Terraform removal for infrastructure controller
**How to categorize this issue?**
/kind technical-debt
/priority 2
/platform aws

**What would you like to be added**:
Project Gardener is using Terraform since beginning to create, manage and delete the underlying infrastructure for Shoot clusters (e.g. VPCs, subnets, nat gateways etc.).

As Terraform was great in the beginning to get started, we today experience more and more issues with it. 
For example:
- Less control over the `Infrastructure` creation and deletion process
- Therefore inflexibility to create infrastructure remedies if they are required (depend on upstream Terraform fixes)
- Inability to import resources in the Terraform state in the Gardener context (Terraformer does not support this)
- Terraform provider version incompatibilities
- State lost issues

Therefore we wanted to start dismantling the Terraform usage in this extension and start to adopt provider specific Go SDKs as replacement.

Ideally all Terraform managed resources should be already tracked in the status section of the `Infrastructure` resource.
In case there are resources not tracked in the status then it need to be populated first before the Go SDK clients can take over.

This issue is meant to track progress of the Terraform removal on this extension.

**Why is this needed**:
See above.

/roadmap cloud

### PR DESCRIPTION
**Pull request by MartinWeindel**
**How to categorize this PR?**

/area control-plane
/kind enhancement
/platform aws

**What this PR does / why we need it**:
Reconciliation and deletion of infrastructure resources are managed directly with a graph flow and using AWS SDK client API instead of using Terraformer.
Migration for existing infrastructure resources is implemented, too.

**Which issue(s) this PR fixes**:
Fixes #457

**Special notes for your reviewer**:
Currently, to activate the flow based infra management, the shoot or the infrastructure resource must be annotated with
`aws.provider.extensions.gardener.cloud/use-flow=true`

**Release note**:

```feature operator
Flow-based infrastructure reconciliation without Terraformer
```

### CODE REVIEW

**Comment by MartinWeindel** [comment]
Regarding the failing verify step: depends on #602 and an update of the verify image to `golang:1.19.x`

File: Makefile [hunk 1]
```diff
@@ -43,8 +43,6 @@ SECRET_ACCESS_KEY_FILE := .kube-secrets/aws/secret_access_key.secret
 #########################################
 
 TOOLS_DIR := hack/tools
-# temporarily until Gardener version has been revendored >= v1.54
-GOLANGCI_LINT_VERSION := v1.49.0
```
Comment (kon-angelo) [inline Makefile:47]: Meanwhile I think that the newest vendored version > 1.50 and we wouldn't need this.

Comment (MartinWeindel) [reply Makefile:47]: It's been removed here. What's your point?

**Comment by MartinWeindel** [comment]
rebased to resolve merge conflicts

File: pkg/controller/infrastructure/infraflow/persistent_state.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+import (
+	"fmt"
+
+	"github.com/gardener/gardener-extension-provider-aws/pkg/controller/infrastructure/infraflow/shared"
+	"k8s.io/apimachinery/pkg/util/json"
+)
+
+const (
+	// PersistentStateVersion is the current version used for persisting the state.
+	PersistentStateVersion = "1.0"
```
Comment (kon-angelo) [inline pkg/controller/infrastructure/infraflow/persistent_state.go:26]: maybe `v1`

File: pkg/aws/client/updater.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+func (u *updater) UpdateRouteTable(ctx context.Context, log logr.Logger, desired, current *RouteTable, controlledCidrBlocks ...string) (modified bool, err error) {
+outerDelete:
+	for _, cr := range current.Routes {
+		for _, dr := range desired.Routes {
+			if reflect.DeepEqual(cr, dr) {
+				continue outerDelete
+			}
+		}
+		if cr.GatewayId != nil && *cr.GatewayId == "local" {
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/aws/client/updater.go:117]: The order of the routes is not guaranteed and the end result would be equivalent.

File: pkg/aws/client/utils.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+func copyArray[T any](array []T) []T {
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/aws/client/utils.go]: done

File: pkg/controller/infrastructure/actuator.go [hunk 5]
```diff
@@ -87,3 +95,14 @@ func generateTerraformerEnvVars(secretRef corev1.SecretReference) []corev1.EnvVa
...
+func (a *actuator) addErrorCodes(err error) error {
+	if err == nil {
+		return nil
+	}
+	if msg := err.Error(); strings.Contains(msg, "AuthFailure:") ||
+		strings.Contains(msg, "InvalidClientTokenId:") {
+		return helper.NewErrorWithCodes(err, gardencorev1beta1.ErrorInfraUnauthorized)
+	}
+	return err
+}
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/actuator.go]: using util.DetermineError instead

File: pkg/controller/infrastructure/actuator_reconcile.go [hunk 6]
```diff
@@ -51,11 +67,232 @@ func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, infrastructur
...
+	oldState *infraflow.PersistentState) error {
+	log.Info("reconcileWithFlow")
+
+	flowContext, err := a.createFlowContext(ctx, log, infrastructure, oldState)
+	if err != nil {
+		return err
+	}
+	if err = flowContext.Reconcile(ctx); err != nil {
+		_ = flowContext.PersistState(ctx, true)
+		return a.addErrorCodes(err)
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/actuator_reconcile.go]: done

File: pkg/controller/infrastructure/infraflow/persistent_state.go (hunk 2 above)
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/persistent_state.go]: using TypeObject with kind and apiVersion instead. I'm using now same the apiVersion as the other apis types: `aws.provider.extensions.gardener.cloud/v1alpha1`

**Comment by MartinWeindel** [comment]
rebased to resolve merge conflicts

**Comment by kon-angelo** [comment]
/needs rebase

otherwise lgtm

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Pull request merged**